### PR TITLE
[MAINT] Format and simplify tests in root

### PR DESCRIPTION
**Pull request by Remi-Gau**
Relates #2528

Changes proposed in this pull request:

- [x] update files include in pre-commit config
- [x] update files excluding in pre-commit
- [x] pre-commit run -a
- [x] fix remaining flake8 issues
- [x] remove dead code flagged by vulture
- [x] apply refactoring from sourcery
- [x] fix remaining f strings
- [x] update flake8 / isort
- [ ] simplify, refactor, parametrize, fixturize tests

### CODE REVIEW

**Comment by Remi-Gau** [comment]
very early PR - just need to start running CI on it
no review needed for now

**Comment by Remi-Gau** [comment]
Still not ready for review

**Comment by ymzayek** [comment]
I noticed that there is a `nilearn/_utils/glm.py` with functions tested in `nilearn/glm/tests/test_utils.py`. Would it make sense to move `glm.py` to a new `_utils.py` under the glm module? First need to check if those functions are used elsewhere, then it could also make sense to move the tests from the glm module to `nilearn/_utils/tests`.

**Comment by Remi-Gau** [comment]
> I noticed that there is a `nilearn/_utils/glm.py` with functions tested in `nilearn/glm/tests/test_utils.py`. Would it make sense to move `glm.py` to a new `_utils.py` under the glm module? First need to check if those functions are used elsewhere, then it could also make sense to move the tests from the glm module to `nilearn/_utils/tests`.

let me have a look, but otherwise I agree that this can be moved.

**Comment by Remi-Gau** [comment]
Ok so except `_check_run_sample_masks` that is used by by nilearn/signal.py, I think all the other functions are used only in the glm folder.

So I think that justify moving this file.

Maybe in another PR though, no? (I think I am way passed the average recommended number of line changes in PR already)

File: nilearn/tests/test_masking.py [hunk 1]
```diff
@@ -1,371 +1,483 @@
...
-    # veriy that 4D masks are rejected
-    mask_img_4d = Nifti1Image(np.ones((40, 40, 40, 2)), np.eye(4))
-    with pytest.raises(DimensionError, match=_TEST_DIM_ERROR_MSG % "4D"):
-        masking.apply_mask(data_img, mask_img_4d)
 
-    # Check that 3D data is accepted
-    data_3d = Nifti1Image(np.arange(27, dtype="int32").reshape((3, 3, 3)),
-                          np.eye(4))
-    mask_data_3d = np.zeros((3, 3, 3))
+def test_apply_mask_3D_data_is_accepted():
```
Comment (NicolasGensollen) [inline nilearn/tests/test_masking.py:324]:
```suggestion
def test_apply_mask_3d_data_is_accepted():
```

File: nilearn/tests/test_masking.py [hunk 2]
```diff
@@ -1,371 +1,483 @@
...
-            t = unmask([masked3D], filename, order="F")
-            t = [get_data(t_) for t_ in t]
-            assert isinstance(t, list)
-            assert t[0].ndim == 3
-            assert not t[0].flags["C_CONTIGUOUS"]
-            assert t[0].flags["F_CONTIGUOUS"]
-            assert_array_equal(t[0], unmasked3D)
+
+@pytest.mark.parametrize("create_files", [False, True])
+def test_unmask_3D(create_files):
```
Comment (NicolasGensollen) [inline nilearn/tests/test_masking.py:412]:
```suggestion
def test_unmask_3d(create_files):
```

File: nilearn/tests/test_masking.py [hunk 3]
```diff
@@ -445,126 +544,219 @@ def test_intersect_masks():
...
+    connected=True (which is by default)
+    """
+    mask_a_img_change_dtype = Nifti1Image(
+        get_data(mask_a_img).astype(">f8"), affine=mask_a_img.affine
+    )
+    mask_b_img_change_dtype = Nifti1Image(
+        get_data(mask_b_img).astype(">f8"), affine=mask_b_img.affine
+    )
+    mask_ab_change_type = intersect_masks(
+        [mask_a_img_change_dtype, mask_b_img_change_dtype], threshold=1.0
```
Comment (NicolasGensollen) [inline nilearn/tests/test_masking.py:653]: Maybe we can remove the lines above and factorize a bit :

```suggestion
        [Nifti1Image(get_data(mask).astype(">8"), affine=mask.affine) for mask in (mask_a_img, mask_b_img)], threshold=1.0
```

**Review by NicolasGensollen — COMMENTED** [review]
Thanks @Remi-Gau ! Looking good ! 
I didn't have the time to look at `test_signal.py` (I'll try to take a look later today).

File: nilearn/glm/tests/test_design_matrix.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    assert_almost_equal(drift_terms_only, cosine_drift)
+
+
+def test_create_cosine_drift_terms_warnings():
+    """Testing cosine filter interface and output."""
+    # fmriprep high pass cutoff is 128s, it's around 0.008 hz
+    t_r, high_pass = 2.5, 0.008
+    signals, _, confounds = generate_signals(
+        n_features=41, n_confounds=5, length=45
+    )
```
Comment (ymzayek) [inline nilearn/glm/tests/test_design_matrix.py:44]: If you foresee more test functions as this and the one above using the same values, could move into a fixture

File: nilearn/tests/test_masking.py [hunk 5]
```diff
@@ -1,371 +1,483 @@
...
+    if standardize_confounds:
+        conf = StandardScaler(with_std=False).fit_transform(conf)
+    cov_mat = _cov_conf(tseries, conf)
+    return np.sum(np.abs(cov_mat))
 
-_TEST_DIM_ERROR_MSG = ("Input data has incompatible dimensionality: "
-                       "Expected dimension is 3D and you provided "
-                       "a %s image")
 
 def _simu_img():
```
Comment (ymzayek) [inline nilearn/tests/test_masking.py:60]: I think it'd be good to seed the random state in this function

File: nilearn/tests/test_masking.py [hunk 6]
```diff
@@ -1,371 +1,483 @@
...
+):
+    """See Issue #2584
+    Code from @pbellec
 
-def test_confounds_standardization():
-    # Tests for confounds standardization
-    # See Issue #2584
-    # Code from @pbellec
-    #
+    Confounds are explicitly standardized
```
Comment (ymzayek) [inline nilearn/tests/test_masking.py:88]: Maybe switch the header and body of this docstring

File: nilearn/tests/test_masking.py [hunk 7]
```diff
@@ -1,371 +1,483 @@
...
+        )
+        < expected * eps
+    )
+
+
+@pytest.mark.parametrize("standardize_signal", [False, "zscore", "psc"])
+def test_confounds_standardization_confounds_are_not_standarized(
+    standardize_signal,
+):
+    """See Issue #2584
```
Comment (ymzayek) [inline nilearn/tests/test_masking.py:106]: same comment about the docstring

File: nilearn/tests/test_masking.py [hunk 8]
```diff
@@ -1,371 +1,483 @@
...
-    tseries1 = masker1.transform(img, confounds=[hv_confounds, conf])
-    masker2 = NiftiMasker(standardize=True, detrend=False,
-                          high_variance_confounds=True,
-                          mask_img=mask).fit()
-    tseries2 = masker2.transform(img, confounds=conf)
-    np.testing.assert_array_equal(tseries1, tseries2)
 
+@pytest.mark.parametrize(
+    "standardize_signal, expected",
+    [(False, 10.0), ("zscore", 1.0), (True, 1.0), ("psc", 10.0)],
```
Comment (ymzayek) [inline nilearn/tests/test_masking.py:80]: I think this should now be:

```suggestion
    [(False, 10.0), ("zscore_sample", 1.0), (True, 1.0), ("psc", 10.0)],
```

See https://github.com/nilearn/nilearn/pull/3474

File: nilearn/tests/test_masking.py [hunk 9]
```diff
@@ -1,371 +1,483 @@
...
-                                 standardize_confounds=True) < 10. * eps)
+    assert (
+        _confounds_regression(
+            standardize_signal=standardize_signal, standardize_confounds=True
+        )
+        < expected * eps
+    )
+
+
+@pytest.mark.parametrize("standardize_signal", [False, "zscore", "psc"])
```
Comment (ymzayek) [inline nilearn/tests/test_masking.py:102]:
```suggestion
@pytest.mark.parametrize("standardize_signal", [False, "zscore_sample", "psc"])
```

File: nilearn/tests/test_masking.py [hunk 10]
```diff
@@ -1,371 +1,483 @@
...
+    mask[6, 6, 6] = 3
+    mask_img = Nifti1Image(mask, AFFINE_EYE)
+    with pytest.raises(ValueError, match="Given mask is not made of 2 values"):
+        apply_mask(Nifti1Image(data, AFFINE_EYE), mask_img)
 
 
-def test_unmask():
-    # A delta in 3D
-    shape = (10, 20, 30, 40)
+def test_unmask_4D():
```
Comment (ymzayek) [inline nilearn/tests/test_masking.py:379]:
```suggestion
def test_unmask_4d():
```
@NicolasGensollen missed one ;)

**Review by ymzayek — COMMENTED** [review]
Partial review up to `test_masking.py`. Will try to finish the rest tomorrow

File: nilearn/tests/__init__.py
Comment (Remi-Gau) [inline nilearn/tests/__init__.py:1]: Given the discussion in https://github.com/nilearn/nilearn/issues/3660

is it premature to remove this or should we say that if the file is removed and the tests can still run then, we might as well remove it.

File: nilearn/tests/test_masking.py (hunk 6 above)
Comment (Remi-Gau) [reply nilearn/tests/test_masking.py:88]: done

File: nilearn/tests/test_masking.py (hunk 7 above)
Comment (Remi-Gau) [reply nilearn/tests/test_masking.py:106]: done

File: nilearn/_version.py
Comment (ymzayek) [inline nilearn/_version.py:1]: How did this pop up here? It's in .gitignore. Doesn't need to be in the source code on github. It will be ncluded in build sdits and wheels

File: nilearn/tests/test_masking.py [hunk 11]
```diff
@@ -1,371 +1,483 @@
...
 def test_compute_brain_mask():
-    img, _ = data_gen.generate_mni_space_img(res=8, random_state=0)
-    brain_mask = compute_brain_mask(img, threshold=.2)
-    gm_mask = compute_brain_mask(img, threshold=.2, mask_type="gm")
-    wm_mask = compute_brain_mask(img, threshold=.2, mask_type="wm")
-    brain_data, gm_data, wm_data = map(get_data, (brain_mask, gm_mask,
-                                                  wm_mask))
+    img, _ = generate_mni_space_img(res=8, random_state=0)
+
+    brain_mask = compute_brain_mask(img, threshold=0.2)
```
[the remark this answers is not in the document]
Comment (Remi-Gau) [reply nilearn/tests/test_masking.py:231]: good idea

File: nilearn/tests/test_masking.py [hunk 12]
```diff
@@ -1,371 +1,483 @@
...
+    assert (
+        _confounds_regression(
+            standardize_signal=standardize_signal, standardize_confounds=True
+        )
+        < expected * eps
+    )
+
+
+@pytest.mark.parametrize("standardize_signal", [False, "zscore", "psc"])
+def test_confounds_standardization_confounds_are_not_standarized(
```
[the remark this answers is not in the document]
Comment (Remi-Gau) [reply nilearn/tests/test_masking.py:103]: done

File: nilearn/tests/test_masking.py [hunk 13]
```diff
@@ -1,371 +1,483 @@
...
-    assert(_confounds_regression(standardize_signal='psc',
-                                 standardize_confounds=True) < 10. * eps)
+    hv_confounds = high_variance_confounds(img)
 
-    # Signal is not standardized
-    # Confounds are not standardized
-    # In this case, the regression should fail...
-    assert(_confounds_regression(standardize_signal=False,
-                                 standardize_confounds=False) > 100)
+    masker1 = NiftiMasker(
```
[the remark this answers is not in the document]
Comment (Remi-Gau) [reply nilearn/tests/test_masking.py:125]: nice

File: nilearn/tests/test_signal.py [hunk 14]
```diff
@@ -470,347 +486,569 @@ def test_clean_kwargs():
...
+    runs[: n_samples // 2] = 0
+    x_detrended = clean(
+        x,
+        confounds=confounds,
+        standardize=False,
+        detrend=True,
+        low_pass=None,
+        high_pass=None,
+        runs=runs,
+        standardize_confounds="zscore",
```
Comment (ymzayek) [inline nilearn/tests/test_signal.py:605]:
```suggestion
        standardize_confounds="zscore_sample",
```

File: nilearn/tests/test_signal.py [hunk 15]
```diff
@@ -470,347 +486,569 @@ def test_clean_kwargs():
...
+            str(filename1),
+            confounds[:, 0:2],
+            str(filename2),
+            confounds[:, 2],
+        ],
+        standardize_confounds="zscore_sample",  # avoid wrong strategy warnings
+    )
+
+
+def test_clean_confounds_from_file_with_confounds_as_pandas_DataFrame():
```
Comment (ymzayek) [inline nilearn/tests/test_signal.py:767]: Same comment as above

File: nilearn/tests/test_masking.py (hunk 3 above)
Comment (Remi-Gau) [reply nilearn/tests/test_masking.py:653]: keeping the comprehension the list comprehension but will keep the intermediate variable to make it a bit easier to parse for humans

**Comment by ymzayek** [comment]
Can you also rebase on main to resolve the conflicts?

File: nilearn/tests/test_masking.py [hunk 16]
```diff
@@ -445,126 +544,219 @@ def test_intersect_masks():
...
+        generate_mni_space_img(res=9, random_state=0)[0],
+        generate_mni_space_img(res=9, random_state=1)[0],
+    ]
+    imgs2 = [
+        generate_mni_space_img(res=9, random_state=2)[0],
+        generate_mni_space_img(res=9, random_state=3)[0],
+    ]
+
+    mask1 = compute_multi_brain_mask(imgs1, threshold=0.2)
+    mask2 = compute_multi_brain_mask(imgs2, threshold=0.2)
```
[the remark this answers is not in the document]
Comment (Remi-Gau) [reply nilearn/tests/test_masking.py:695]: ok maybe this is me still being a green pythonista but I am not sure this will help readibility.

**Comment by Remi-Gau** [comment]
> Can you also rebase on main to resolve the conflicts?

done

File: nilearn/_version.py
Comment (Remi-Gau) [reply nilearn/_version.py:1]: possibility: I installed from `main`, switched branch and made a "commit all" before having rebased and therefore using the new `.gitingore`

Comment (Remi-Gau) [reply nilearn/_version.py:1]: will remove it

File: nilearn/tests/test_signal.py [hunk 17]
```diff
@@ -470,347 +486,569 @@ def test_clean_kwargs():
...
+        standardize="zscore",
+        standardize_confounds="zscore_sample",  # avoid wrong strategy warnings
+    )
+
+    assert_almost_equal(
+        cleaned_signals.var(axis=0), np.ones(cleaned_signals.shape[1])
+    )
+
+
+def test_clean_confounds_from_file():
```
[the remark this answers is not in the document]
Comment (Remi-Gau) [reply nilearn/tests/test_signal.py:713]: not really easy to parametrize in way that won't be ugly AF so will put a loop

[the remark this answers is not in the document]
Comment (ymzayek) [reply nilearn/tests/test_signal.py:710]: Yea figured that could be the case. can also leave as is if it's the most readable

File: nilearn/tests/test_masking.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (NicolasGensollen) [reply nilearn/tests/test_masking.py:687]: No worries. Being readable is way more important than using tricks to reduce the number or lines, so let's forget about this suggestion.

File: nilearn/tests/test_masking.py [hunk 18]
```diff
@@ -580,80 +761,105 @@ def test__extrapolate_out_mask():
...
+            ],
+            [
+                [False, False, False, False, False],
+                [False, False, False, False, False],
+                [False, False, True, False, False],
+                [False, False, False, False, False],
+                [False, False, False, False, False],
+            ],
+        ]
+    )
```
Comment (ymzayek) [inline nilearn/tests/test_masking.py:841]: I think this is equivalent:

```suggestion
    target_mask = np.not_equal(target_data, 0)
```

Comment (Remi-Gau) [reply nilearn/tests/test_masking.py:841]: and only "just" a bit shorter. 🤩

File: nilearn/tests/test_masking.py [hunk 19]
```diff
@@ -1,371 +1,478 @@
...
+    """When confounds are not standardized, the regression should fail.
+
+    See Issue #2584
+    Code from @pbellec
+    """
+    assert (
+        _confounds_regression(
+            standardize_signal=standardize_signal, standardize_confounds=False
+        )
+        > 100
```
Comment (bthirion) [inline nilearn/tests/test_masking.py:116]: Not super clear to me why this indicates a "failure" ?

File: nilearn/tests/test_masking.py [hunk 20]
```diff
@@ -389,53 +507,29 @@ def test_intersect_masks_filename():
...
+@pytest.fixture
+def mask_b():
+    return make_mask_b()
 
-    # Create dummy masks
-    mask_a = np.zeros((4, 4, 1), dtype=bool)
-    mask_a[2:4, 2:4] = 1
-    mask_a_img = Nifti1Image(mask_a.astype("int32"), np.eye(4))
 
-    # +---+---+---+---+
```
Comment (bthirion) [inline nilearn/tests/test_masking.py:410]: Why are you removing this ?

**Review by bthirion — COMMENTED** [review]
LGTM otherwise.

File: nilearn/tests/test_signal.py [hunk 21]
```diff
@@ -469,347 +483,544 @@ def test_clean_kwargs():
...
-    nisignal.clean(signals, detrend=False, standardize=False,
-                   confounds=filename2)
-    nisignal.clean(signals, detrend=False, standardize=False,
-                   confounds=confounds[:, 1])
 
+    cleaned_signals = clean(
+        input_signals,
+        detrend=False,
+        standardize=False,
+        standardize_confounds="zscore_sample",  # avoid wrong strategy warnings
```
Comment (ymzayek) [inline nilearn/tests/test_signal.py:691]: Is this comment referring to the deprecation warning? If so I would remove it because it's not informative and could be misleading

File: nilearn/glm/tests/test_design_matrix.py [hunk 22]
```diff
@@ -0,0 +1,64 @@
+"""Test design matrix creation and manipulation."""
+
+import numpy as np
+import pytest
+
+# Use nisignal here to avoid name collisions
+# (using nilearn.signal is not possible)
```
[the remark this answers is not in the document]
Comment (ymzayek) [reply nilearn/glm/tests/test_design_matrix.py:7]: This does not seem to be resolved

File: nilearn/tests/test_signal.py [hunk 23]
```diff
@@ -103,14 +49,24 @@ def generate_trends(n_features=17, length=41):
 
 
 def generate_signals_plus_trends(n_features=17, n_samples=41):
```
Comment (ymzayek) [inline nilearn/tests/test_signal.py:51]: Probably a note for a separate refactoring PR but this function exists yet several tests re-implement the lines instead of using it

**Comment by Remi-Gau** [comment]
closing this for now as the merge conflict have grown beyond recuperable
